### DISCUSSION TITLE
Standardised / built-in method for namespacing / prefixing template tag variables

### DISCUSSION THREAD

**Discussion by jcogs-design** [ExpressionEngine/ExpressionEngine#3313] [Feature Requests]
**Is your feature request related to a problem? Please describe.**
Naming clashes relating to template tag variables occur. Some add-ons allow for an arbitrary prefix to be added to variable names ([e.g.](https://github.com/litzinger/URL-Helper/blob/master/README.md#configuration)). Others seek to name variables in a way that avoids clashes (e.g. see description of [CT Version 2.0249](https://www.cartthrob.com/docs/pages/change_log/index.html)).

This ad-hoc approach is problematic. Different developers solve it in different ways, and even then there is no guarantee that a variable name clash you encounter will be one covered by such ad-hoc solutions. Further such solutions rely upon the quality of add-on documentation to be effective: the CT one is not mentioned (AFAIK) in documentation at all.

**Describe the solution you'd like**
For EE to introduce a standarised method for namespacing (or prefixing) tag variables - so that if you have a problem you can fix it using the standard method. I can think of two possible approaches:

- support the addition of the add-on method linked to the tag that generates the variable to be linked to as a namespace type call - e.g. if the tag is `{exp:my_addon:my_method}` support optionally the use of `{my_method\my_variable}` in addition to `{my_variable}`
- support the use of the tag method as a prefix - e.g. `my_method_my_variable`

Either way the option could be enabled via a template tag parameter such as `use_alt_naming="yes"` or some such.

**Describe alternatives you've considered**
There are various alternative methods to avoid naming clashes (e.g. use of layouts or embeds etc).

**Teachability, Documentation, Adoption, Migration Strategy**
The change would be required in the `parse_varariables` code - adding in an optional search for the method calling parse_variables and removing the prefix / namespace prior to parsing.

Since the change would be enabled via an optional parameter in the tag, it should be backwards compatible.

- [x] I am capable and would like to work on implementation of this feature if it is considered (but I don't know how much time I'll have in near future - and makes sense to wait until consensus on whether this is a good idea etc. etc...

**Comment by robsonsobral** [comment]
I was thinking about it yesterday.

There's a "tradition" of using `:` as hierarchy separator on tags. The common tags `count`, `total_results` and `no_results` could have an optional syntax prefixed by the method used:

```html
{exp:module:method}
  {if method:count == 1}
    <pre>
  {/if}
    Let's debug!
  {if method:count == method:total_results}
    </pre>
  {/if}
{/exp:module:method}
```

**Comment by jcogs-design** [comment]
Yes - I did think of that, and agree there is a `:` convention that would maybe fit easier in tradition.

My reason for suggesting `method\` or `method_` is that possibly there is less scope for confusion about intent - since both would be easy to strip from front of variable and there is (for the inverse of the reason above) little potential for confusion about what the template designer meant. Especially with the arrival of template modifiers there already are reasons for the variable name to have a `:suffix` form - and if we add in fluid and grid fields also reasons for it to have a `collection_variable:` prefix - so I wondered if it would get potentially confusing to have yet another `:` lurking around - harder to read / interpret by the designer and harder to process for the `parse_variables` method.

**Comment by robsonsobral** [comment]
I'm sorry, but I think that to have conditional standards is more confusing. Like the last big changes to Sass and Husky showed, to add a new behavior should be the last option. If it could be avoided...

However, I know nothing about the new modifiers you mentioned. Maybe your suggestion is already the "less worst" considering them.

**Comment by jcogs-design** [comment]
Surely - I agree changes should be minimised as a first option.

However I think adding yet another use for the `:` in variable naming would be more confusing.  We already have `grid` and `field` options that prefix the variable name separated by a `:` - e.g. 
```
{layout:my_variable}
{my_grid:my_variable}
{my_fluid_field:my_variable}
```
If we add-in [variables modifiers](https://docs.expressionengine.com/latest/templates/variable-modifiers.html#variable-modifiers) we have potentially two colons in the name:
```
{layout:my_variable:my_modifier}
{my_grid:my_variable:my_modifier}
{my_fluid_field:my_variable:my_modifier}
```
If we than consider adding in namespacing the variable we have multple options (depending on whether we are namspacing the grid/fluid field or the variable itself...
```
{layout:my_namespace:my_variable:my_modifier}
{my_grid:my_namespace:my_variable:my_modifier}
{my_fluid_field:my_namespace:my_variable:my_modifier}
{my_namespace:my_grid:my_variable:my_modifier}
{my_namespace:my_fluid_field:my_variable:my_modifier}
{my_other_namespace:my_grid:my_namespace:my_variable:my_modifier}
{my_other_namespace:my_fluid_field:my_namespace:my_variable:my_modifier}
``` 
Trying to parse that is not trivial.
Using the `\` option makes this a little more understandable... (perhaps)
```
{layout:my_namespace\my_variable:my_modifier}
{my_grid:my_namespace\my_variable:my_modifier}
{my_fluid_field:my_namespace\my_variable:my_modifier}
{my_namespace\my_grid:my_variable:my_modifier}
{my_namespace\my_fluid_field:my_variable:my_modifier}
{my_other_namespace\my_grid:my_namespace\my_variable:my_modifier}
{my_other_namespace\my_fluid_field:my_namespace\my_variable:my_modifier}
``` 
Having a common / standard method for namespacing variables would be potentially very helpful - if applied at the level of `parse_variables` it would potentially automatically gift namespacing to existing addons that use `parse_variables` - and cut through the 'synonym for no_results' and home-brewed prefix methods that exist and / or might be created.

**Comment by robsonsobral** [comment]
I see... Let's split into 2!

1. in the case of `count` and `total_results`, as grid already does, we can add `module:` as an alternative syntax. I guess we can implement this easier than the rest.
2. all the other situations looks so cluttered to me. I don't know yet what to think of them. I guess using only one "separator" looks a little better. Hmmmmnnnnn...
